sparse/src/KokkosSparse_spmv.hpp

```diff
@@ -859,459 +90,422 @@ void spmv(const ExecutionSpace& space,
+      typename XVector::device_type,
+      Kokkos::MemoryTraits<Kokkos::Unmanaged | Kokkos::RandomAccess>>;
+
+  using YVector_Internal = Kokkos::View<
+      typename YVector::non_const_data_type,
+      typename KokkosKernels::Impl::GetUnifiedLayout<YVector>::array_layout,
+      typename YVector::device_type, Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+
+  XVector_Internal x_i(x);
+  YVector_Internal y_i(y);
+
+  bool useNative = is_spmv_algorithm_native(handle->get_algorithm());
+  // Also use the native algorithm if SPMV_FAST_SETUP was selected and
+  // rocSPARSE is the possible TPL to use. Native is faster in this case.
+#ifdef KOKKOSKERNELS_ENABLE_TPL_ROCSPARSE
+  if (handle->get_algorithm() == SPMV_FAST_SETUP &&
+      std::is_same_v<ExecutionSpace, Kokkos::HIP>)
+    useNative = true;
+#endif
+
+  // Now call the proper implementation depending on isBSR and the rank of X/Y
+  if constexpr (!isBSR) {
+    if constexpr (XVector::rank() == 1) {
+/////////////////
+// CRS, rank 1 //
+/////////////////
 #ifdef KOKKOSKERNELS_ENABLE_TPL_CUSPARSE
-  // cuSPARSE does not support the modes (C), (T), (H)
-  if (std::is_same<typename AMatrix_Internal::memory_space,
-                   Kokkos::CudaSpace>::value ||
-      std::is_same<typename AMatrix_Internal::memory_space,
-                   Kokkos::CudaUVMSpace>::value) {
-    useFallback = useFallback || (mode[0] != NoTranspose[0]);
-  }
+      // cuSPARSE does not support the conjugate mode (C)
+      if constexpr (std::is_same_v<typename AMatrix_Internal::memory_space,
+                                   Kokkos::CudaSpace> ||
+                    std::is_same_v<typename AMatrix_Internal::memory_space,
+                                   Kokkos::CudaUVMSpace>) {
+        useNative = useNative || (mode[0] == Conjugate[0]);
+      }
+      // cuSPARSE 12 requires that the output (y) vector is 16-byte aligned for
+      // all scalar types
+#if defined(CUSPARSE_VER_MAJOR) && (CUSPARSE_VER_MAJOR == 12)
+      uintptr_t yptr = uintptr_t((void*)y.data());
+      if (yptr % 16 != 0) useNative = true;
+#endif
+#endif
+
+#ifdef KOKKOSKERNELS_ENABLE_TPL_ROCSPARSE
+      if (std::is_same<typename AMatrix_Internal::memory_space,
+                       Kokkos::HIPSpace>::value) {
+        useNative = useNative || (mode[0] != NoTranspose[0]);
+      }
 #endif
 
 #ifdef KOKKOSKERNELS_ENABLE_TPL_MKL
-  if (std::is_same<typename AMatrix_Internal::memory_space,
-                   Kokkos::HostSpace>::value) {
-    useFallback = useFallback || (mode[0] == Conjugate[0]);
-  }
+      if (std::is_same_v<typename AMatrix_Internal::memory_space,
+                         Kokkos::HostSpace>) {
+        useNative = useNative || (mode[0] == Conjugate[0]);
+      }
+#ifdef KOKKOS_ENABLE_SYCL
+      if (std::is_same_v<typename AMatrix_Internal::memory_space,
+                         Kokkos::Experimental::SYCLDeviceUSMSpace>) {
+        useNative = useNative || (mode[0] == Conjugate[0]);
```

Comment:
Actually the conjugate mode is supported by oneMKL see here: https://www.intel.com/content/www/us/en/docs/onemkl/developer-reference-dpcpp/2024-0/oneapi-mkl-sparse-gemv.html
We may want to keep this for another PR but we should note that this is something that should be offloaded to the TPL just fine.

Reply:
@lucbv I see normal, transpose and conjugate transpose (what we call N, T and H) but not conjugate non-transpose (what we call C, which is when we fall back)

Reply:
Yeah, I thought "C" meant transpose conjugate so "H" but should have read things more carefully.